### PR TITLE
🐛Destination-dynamodb: enforce ssl connection

### LINKED ISSUE #16283: Add ability to enforce SSL in DynamoDB Destination connector
## Tell us about the problem you're trying to solve
DynamoDB does not _always_ enforce SSL 

## Describe the solution you’d like
Implement configuration for adding `ssl` parameter for DynamoDB destination

## Describe the alternative you’ve considered or used
A clear and concise description of any alternative solutions or features you've considered or are using today.

## Additional context
[Client-side Encryption DynamoDB](https://github.com/aws/aws-dynamodb-encryption-java/)

## Are you willing to submit a PR?

Remove this with your answer :-)

### PR DESCRIPTION
**Pull request by etsybaev**
## What
Users are able to use a non-secure connection

## How
Updated check method to enforce SSL connection.

Selection_032

## 🚨 User Impact 🚨
No breaking changes expected

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestination.java
Comment (suhomud) [inline airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestination.java:37]: Minor proposal to specify via method name what we are checking. 

  ```
  if (isNotSsl(dynamodbDestinationConfig.getEndpoint())) {
        return new AirbyteConnectionStatus()
            .withStatus(AirbyteConnectionStatus.Status.FAILED)
            .withMessage(NON_SECURE_URL_ERR_MSG);
    }
```
method:
```
  public boolean isNotSsl(String endpoint) {
    return !endpoint.isBlank() &&
        new URL(endpoint).getProtocol().equals(Protocol.HTTP.toString());
  }
```

Comment (etsybaev) [reply airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestination.java]: Good catch, updated. Thanks

**Review by suhomud — APPROVED** [review]
LGTM

**Comment by etsybaev** [comment]
Selection_033

**Review by grishick — CHANGES_REQUESTED** [review]
If I am reading the code correctly, this change will always enforce HTTPS for DynamoDB. However, we only need this to happen in Airbyte Cloud and we want to continue allowing non-SSL connections in Open Source. @ryankfu could you walk @etsybaev through how to introduce strict encrypt variant to an existing destination connector w/o creating a new image and how to add the check only in Airbyte Cloud scope (what you did for destination-S3)?

**Comment by ryankfu** [comment]
Hi @etsybaev, in case you weren't aware there's a PR that captures a very similar change [here](https://github.com/airbytehq/airbyte/pull/17340/files). Basically the main pieces that would help to separate this functionality so that it only affects cloud is to:
* create a new `DynamoDestinationRunner` and `DynamoDestinationStrictEncrypt` in the same package as `DynamoDestination`
* update dynamo destination's `build.gradle` to point to the `DynamoDestinationRunner`

If you have any questions regarding the work, let's set up a quick 1:1 to knowledge share. For most context the `Runner` class will act similar to a case scenario where it will run a different class based on the environment variable

Also note that you'd probably want to create a new `StrictEncryptTest` so that it only tests for this behavior in an isolated manner

File: airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncrypt.java
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncrypt.java:34]: Nit: this seems redundant since you call `super.check(config)` outside the try/catch block since that already does this validation method

A suggestion would be to have the code like

```java
@Override
  public AirbyteConnectionStatus check(final JsonNode config) {
    final S3DestinationConfig destinationConfig = this.configFactory.getS3DestinationConfig(config, super.storageProvider());

    // enforce ssl connection
    if (!DynamodbChecker.testCustomEndpointSecured(destinationConfig.getEndpoint())) {
      return new AirbyteConnectionStatus()
                .withStatus(Status.FAILED)
          .withMessage("NON_SECURE_URL_ERR_MSG");
    }
    return super.check(config);
  }
```

The try/catch block isn't necessary since the method `testCustomEndpointSecured` is defined by you and it doesn't throw any `Exceptions` and `DynamodbDestination` already covers the try/catch within it's `check()` operation

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java:17]: Unless I'm mistaken you can probably reuse the `NON_SECURE_URL_ERR_MSG` variable by making it `protected` in `DynamodbDestinationStrictEncrypt`, this creates a single source of truth so code doesn't deviate over time and easier to maintain

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java:23]: This is good, can we also just have a sanity check test for `https` variants return an `AirbyteConnectionStatus.SUCCEEDED`, this captures the scenario where maybe the custom endpoint gets augmented and we fail on `https` in the future

**Review by ryankfu — CHANGES_REQUESTED** [review]
minor code reusability changes but overall looks good

File: airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncrypt.java
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-dynamodb/src/main/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncrypt.java]: Good catch, updated, thanks!

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java]: Updated, thanks

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java]: Added one more test. Thanks

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (ryankfu) [inline airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java:35]: nit: type for `fails`

File: build.gradle
Comment (ryankfu) [inline build.gradle:159]: How come you've commented out the `prettier()` method? presume this is simply pretty printing but regardless doesn't seem relevant to this PR

**Review by ryankfu — APPROVED** [review]
Overall looks good, there's a minor nit and a question regarding a `prettier()` method getting commented out in `build.gradle`

File: airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-dynamodb/src/test/java/io/airbyte/integrations/destination/dynamodb/DynamodbDestinationStrictEncryptTest.java]: Updated, thanks

File: build.gradle
Comment (etsybaev) [reply build.gradle]: Accidentally committed. Already rolled back.
How this could come? There was a version bumping change in check-style some time ago. Since then some linux users started seeing issues with importing and refreshing project in IntellijIdea but still works when executing build from console. I raised that issue but it seems like the author was busy with other tasks and wasn't going to switch back to his changes. I tried to check it in my free time and found that the issue comes from "prettier" and then stuck. Also found a few similar issues raised on the internet but without answers. So since then every time I have to comment it while working with code and then uncomment from time to time to run "./gradlew --no-daemon format --scan" command to format code and then comment it back. 
Here is what the issue looks like for me in IntellijIdea when I refresh gradle project with uncommented Prettier
Selection_035

**Review by grishick — CHANGES_REQUESTED** [review]
@etsybaev now that you've made the changes to enforce SSL only in Cloud you also need to bump the version again and publish the newer version before merging

**Review by grishick — APPROVED** [review]
Approved assuming you'll bump the connector version and publish ;)

**Comment by etsybaev** [comment]
Selection_038

**Pull request merged**
